pkg/manager/member/tiproxy_member_manager.go

```diff
@@ -0,0 +1,656 @@
+		if err != nil {
+			return err
+		}
+		err = m.deps.StatefulSetControl.CreateStatefulSet(tc, newSts)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	// Scaling takes precedence over upgrading because:
+	// - if a pod fails in the upgrading, users may want to delete it or add
+	//   new replicas
+	// - it's ok to scale in the middle of upgrading (in statefulset controller
+	//   scaling takes precedence over upgrading too)
+	if err := m.scaler.Scale(tc, oldStatefulSet, newSts); err != nil {
+		return err
+	}
+
+	if !templateEqual(newSts, oldStatefulSet) || tc.Status.TiProxy.Phase == v1alpha1.UpgradePhase {
+		if err := m.upgrader.Upgrade(tc, oldStatefulSet, newSts); err != nil {
+			return err
+		}
+	}
+
+	return mngerutils.UpdateStatefulSetWithPrecheck(m.deps, tc, "FailedUpdateTiProxySTS", newSts, oldStatefulSet)
+}
+
+func (m *tiproxyMemberManager) syncStatus(tc *v1alpha1.TidbCluster, sts *apps.StatefulSet) error {
+	if sts == nil {
+		// skip if not created yet
+		return nil
+	}
+
+	tc.Status.TiProxy.StatefulSet = &sts.Status
+	upgrading, err := m.statefulSetIsUpgradingFn(m.deps.PodLister, m.deps.PDControl, sts, tc)
+	if err != nil {
+		tc.Status.TiProxy.Synced = false
+		return err
+	}
+	if upgrading {
+		tc.Status.TiProxy.Phase = v1alpha1.UpgradePhase
+	} else {
+		tc.Status.TiProxy.Phase = v1alpha1.NormalPhase
+	}
+
+	replicas := 0
+	sameConfigProxy := true
+	pods := helper.GetPodOrdinals(tc.Status.TiProxy.StatefulSet.Replicas, sts)
+	for id := range pods {
+		podHealthy, err := m.deps.ProxyControl.IsPodHealthy(tc, int32(id))
+		if err != nil {
+			return err
+		}
+		if podHealthy {
+			replicas++
+		}
+		cfg, err := m.deps.ProxyControl.GetConfigProxy(tc, pods.UnsortedList()[0])
+		if err != nil {
+			return err
+		}
+		if sameConfigProxy && tc.Spec.TiProxy.Proxy.MaxConnections != cfg.MaxConnections || tc.Spec.TiProxy.Proxy.TCPKeepAlive != cfg.TCPKeepAlive {
+			sameConfigProxy = false
+			tc.Status.TiProxy.Proxy = *cfg
+			if err := m.deps.ProxyControl.SetConfigProxy(tc, pods.UnsortedList()[0], tc.Spec.TiProxy.Proxy); err != nil {
+				return err
+			}
+		}
+	}
+
+	tc.Status.TiProxy.Synced = replicas == int(tc.TiProxyDeployDesiredReplicas()) && sameConfigProxy
+	return nil
+}
+
+func (m *tiproxyMemberManager) syncProxyService(tc *v1alpha1.TidbCluster, peer bool) error {
+	svcLabel := labelTiProxy(tc)
+	newSvc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace:       tc.GetNamespace(),
+			OwnerReferences: []metav1.OwnerReference{controller.GetOwnerRef(tc)},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector:                 svcLabel,
+			PublishNotReadyAddresses: true,
+		},
+	}
+	if peer {
+		newSvc.ObjectMeta.Name = controller.TiProxyPeerMemberName(tc.GetName())
+		newSvc.ObjectMeta.Labels = svcLabel.Copy().UsedByPeer()
+		newSvc.Spec.Type = corev1.ServiceTypeClusterIP
+		newSvc.Spec.ClusterIP = "None"
+		newSvc.Spec.Ports = append(newSvc.Spec.Ports,
```

Comment:
Will it support some service config in TidbCluster CR later?

Reply:
No. It is fixed.